File: working-effectively-with-legacycode-java/src/test/java/com/uj/extractAndOverrideFactoryMethod/WorkflowEngineTest.java
package com.uj.extractAndOverrideFactoryMethod;

import org.junit.Test;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

/**
 * Created by jehan on 1/4/2015.
 */
public class WorkflowEngineTest {
    @Test
    public void realRunisTrue() throws Exception {
        WorkflowEngine engine = new WorkflowEngine();
        assertTrue(engine.realRun());
    }

    @Test
    public void realRunisFalse() throws Exception {
        WorkflowEngine engine = new TestWorkflowEngine();
        assertFalse(engine.realRun());
    }




    class FakeTransactionManager extends TransactionManager {
        public FakeTransactionManager(Reader reader, Persister persister) {
            super(reader, persister);
        }

        @Override
        public boolean doSomething() {
            return false;
        }
    }

    class TestWorkflowEngine extends WorkflowEngine {
        @Override
        protected TransactionManager makeTransactionManager() {
            return new FakeTransactionManager(null, null);
        }
    }

}

File: working-effectively-with-legacycode-cc++/cpp/biz/test/ExtractAndOverrideGetter/WorkflowEngineTest.cpp
/*
 * WorkflowEngineTest.cpp
 *
 *  Created on: Dec 30, 2014
 *      Author: jehan
 */

#include "gtest/gtest.h"
#include "ExtractAndOverrideGetter/WorkflowEngine.h"

using namespace ExtractAndOverrideGetter;


class FakeTransactionManager : public TransactionManager {
public:
	bool doSomething() {
		return false;
	}
};

class TestWorkflowEngine : public WorkflowEngine {
private:
	FakeTransactionManager fake;
public:
	TransactionManager* getTransactionManager() {
		return &fake;
	}
};



TEST(WorkflowEngineTest, runWithFakeTransactionManager) {
	TestWorkflowEngine engine;
	ASSERT_FALSE(engine.realRun());
}

